### DISCUSSION TITLE
Get parameters update when new YAMCS gets new tm data

### DISCUSSION THREAD

**Discussion by codenamefrance** [yamcs/yamcs#962] [Q&A]
I'm trying to create a python script to get realtime update when YAMCS receives a new telemetry packet; i want the script to be able to read tm parameters and to use them in other computations.
My idea is to use websockets, but it is not so clear to me which parameters to send to yamcs in order to get this kind of update.

**Comment by xpromache** [comment]
Have you seen these examples:
https://docs.yamcs.org/python-yamcs-client/examples/

parameter_subscription.py is maybe what you need.

**Reply by codenamefrance** [reply]
parameter_subscription.py uses yamcs python library. Instead, I'd like to use a websocket (in case I'd want to do a similar thing in javascript too). My problem is that I can't figure out which are the topics to subscribe in order to get this kind of data.

**Reply by fqqb** [reply]
Subscribing to parameter updates is done with a topic called "parameters": https://docs.yamcs.org/yamcs-http-api/processing/subscribe-parameters/

Even if you don't use it, the Python Yamcs Client library can still be useful to figure out how everything works. Or use your browser's developer tools.

**Comment by codenamefrance** [comment]
I tried to write down this code: 
`
import asyncio
import websockets

async def connect_and_listen():
    url = "ws://localhost:8090/api/websocket"
    
    async with websockets.connect(url) as websocket:
        # Invia un messaggio al server
        message = {
            "type": "parameters", 
            "options": {
                "instance": "node1",
                "processor": "realtime",
            }
        }
        await websocket.send(str(message))
        print(f"Messaggio inviato: {message}")

        while True:
            try:
                response = await websocket.recv()
                print(f"Messaggio ricevuto: {response}")
            except websockets.ConnectionClosed:
                print("Connessione chiusa")
                break

asyncio.get_event_loop().run_until_complete(connect_and_listen())
`

and I'm able to get this response:
`Messaggio ricevuto: {
  "type": "parameters",
  "call": 78,
  "seq": 1,
  "data": {
    "@type": "/yamcs.protobuf.processing.SubscribeParametersData"
  }`

However, I'm not able to get the actual data, even if I'm sending it to yamcs(and I cant visualize it from there).

**Reply by fqqb** [reply]
You need to be explicit what parameters you subscribe to. Otherwise it's just an empty subscription.

**Reply by codenamefrance** [reply]
I'm trying putting it in the "id" field, as is written in the docs
` // Parameter identifiers. Each identifier takes the form of
  // a namespace and a name.
  //
  // For Yamcs-native naming only the name field is required and
  // should be the fully qualified name. The namespace is only
  // required when the name represents an alias of that parameter.
  id: NamedObjectId[];`

something like this:
`        message = {
            "type": "parameters", 
            "options": {
                "instance": "node1",
                "processor": "realtime",
                "id": "myproject/gps_alt"
            }
        }`

but it doesn't really work.

**Reply by fqqb** [reply]
The structure of the id field is described on the page that I linked. It's an array of NamedObjectId items.

**Reply by codenamefrance** [reply]
I'm not really sure how this works, I can't get anything to work properly. Could you provide an example?

**Reply by fqqb** [reply]
I don't know your parameters, but for quickstart `id` could be something like this: `[{"name": "/myproject/Latitude"}]`.

**Reply by codenamefrance** [reply]
> I don't know your parameters, but for quickstart `id` could be something like this: `[{"name": "/myproject/Latitude"}]`.

The id field requires an int32 (java error from yamcs)

09:56:44.195 _global [46] WebSocketFrameHandler b48d51c9 Closing channel due to error: com.google.protobuf.InvalidProtocolBufferException: Not an int32 value: [{"namespace":"myproject","name":"gps_alt"}]

**Reply by fqqb** [reply]
The id field at the top level is a number, not the one inside options. You had that part right before.
And your NamedObjectId is not correct either.
